climada/hazard/tc_tracks_synth.py

```diff
@@ -137,25 +162,40 @@ def _one_rnd_walk(track, ens_size, ens_amp0, ens_amp, max_angle, rnd_vec):
     """
     ens_track = list()
     n_dat = track.time.size
-    xy_ini = ens_amp0 * (rnd_vec[:2 * ens_size].reshape((2, ens_size)) - 0.5)
-    tmp_ang = np.cumsum(2 * max_angle * rnd_vec[2 * ens_size:] - max_angle)
-    coord_xy = np.empty((2, ens_size * n_dat))
-    coord_xy[0] = np.cumsum(ens_amp * np.sin(tmp_ang))
-    coord_xy[1] = np.cumsum(ens_amp * np.cos(tmp_ang))
+    n_seg = n_dat - 1
+    xy_ini = max_shift_ini * (2 * rnd_vec[:2 * ens_size].reshape((2, ens_size)) - 1)
+    dt = np.unique(track['time_step'])[0]
 
     ens_track.append(track)
     for i_ens in range(ens_size):
         i_track = track.copy(True)
 
-        d_xy = coord_xy[:, i_ens * n_dat: (i_ens + 1) * n_dat] - \
-            np.expand_dims(coord_xy[:, i_ens * n_dat], axis=1)
-        # change sign of latitude change for southern hemishpere:
-        d_xy = np.sign(track.lat.values[0]) * d_xy
-
-        d_lat_lon = d_xy + np.expand_dims(xy_ini[:, i_ens], axis=1)
-
-        i_track.lon.values = i_track.lon.values + d_lat_lon[0, :]
-        i_track.lat.values = i_track.lat.values + d_lat_lon[1, :]
+        # angular perturbation
+        i_start_ang = 2 * ens_size + i_ens * n_seg
+        i_in_ang = (i_start_ang, i_start_ang + track.time.size - 1)
+        ang_pert = dt * np.degrees(max_ddir * (2 * rnd_vec[i_in_ang[0]:i_in_ang[1]] - 1))
+        ang_pert_cum = np.cumsum(ang_pert)
+        bearings = _get_bearing_angle(i_track.lon.values, i_track.lat.values)
+        angular_dist = _get_angular_distance(i_track.lon.values, i_track.lat.values)
+
+        # for translational speed perturbation:
+        i_start_trans = 2 * ens_size + ens_size * n_seg + i_ens * n_seg
+        i_in_trans = (i_start_trans, i_start_trans + track.time.size - 1)
+        trans_pert = 1 + max_dspeed_rel * (2 * rnd_vec[i_in_trans[0]:i_in_trans[1]] - 1)
+
+        # new lon/lat
+        new_lon = np.zeros_like(i_track.lon.values)
+        new_lat = np.zeros_like(i_track.lat.values)
+        new_lon[0] = i_track.lon.values[0] + xy_ini[0, i_ens]
+        new_lat[0] = i_track.lat.values[0] + xy_ini[1, i_ens]
+        for i in range(0, len(new_lon) - 1):
+            new_lon[i + 1], new_lat[i + 1] = \
+                _get_destination_points(new_lon[i], new_lat[i],
+                                        bearings[i] + ang_pert_cum[i],
+                                        trans_pert[i] * angular_dist[i])
+
+        i_track.lon.values = new_lon
+        i_track.lat.values = new_lat
```

Comment:
This part of the code is very hard to read. Many obscure variables and hard-coded indices.